### PR TITLE
Feature: `i18n` module

### PR DESCRIPTION
**Pull request by bystam**
## What?
- Implement a whole new API for the translations lib

## Why?
- We are copy-pasting these files across multiple services, let's try and imagine what it could look like if we centralise it

------
Here's the README

# i18n

`i18n` is a module that provides an API for getting localized strings based on the locale of the caller.

## Setup
Just add the dependency to maven, and it will auto-configure itself.

## Usage

```kotlin
@GetMapping
fun getContract(): Contract {
    return Contract(
        id = ...,
        // This line will produce the translation for "CONTRACT_DISPLAY_NAME",
        // translated to the language dictated by the `Hedvig-Language` header
        displayName = I18n.get("CONTRACT_DISPLAY_NAME")
    )
}

```

## How it works

The library is set up using auto configuration, that will:
- Start a poller that continuously polls S3 for the most recent strings file
- Add a `jakarta.servlet.Filter` instance that installs a context thread-local that carries the locale
  of the caller, determined by the `Hedvig-Language` header.

Based on this locale thread local, calls such as `I18n["MY_KEY"]` will localize the input key with the locale
of the current HTTP code path.

### Async execution / thread hopping
Sometimes one wants to do async execution, such as when working with GraphQL data loaders. The library ships with
a class called `I18nContextPreservingExecutor`, which is a decorator for `java.util.concurrent.Executor` that
preserves the locale context thread-local across executions. This is needed, or else locale information
is lost on the new thread.

This can be used like this, which is a common pattern we do in some services:
```kotlin
val executor = I18nContextPreservingExecutor(ForkJoinPool.commonPool())
fun <T> supplyAsyncPreservingContext(operation: () -> T): CompletionStage<T> {
    return CompletableFuture.supplyAsync({
        operation()
    }, executor)
}
```

### Patterns for string reuse
Maybe multiple locations in your code wants to use the same translation, or you just want to tuck away hard
coded strings. Here are some simply patterns that can be built on top of this to achieve that.

```kotlin
// You can add extension functions to the I18n type in order to have more high level methods for localization.
fun I18n.welcomePhrase(name: String) = this["WELCOME_PHRASE", mapOf("name" to name)]
// usage
fun getWelcome(member: Member): String = I18n.welcomePhrase(member.firstName)

// You can even do something more crazy if you wish to have namespacing
// placing translations in a namespace object
object QuoteLocalizedStrings {
    fun underwritingGuidelineFailure(guideline: String): String = I18n.get("UNDERWRITING_GUIDELINE_FAILURE_$guideline")
}

val I18n.quotes: QuoteLocalizedStrings get() = QuoteLocalizedStrings
// usage
val errorMessage = I18n.quotes.underwritingGuidelineFailure(guideline)
```

### CODE REVIEW

**Review by alebedev — APPROVED** [review]
Great work!

**Review by mightyost — APPROVED** [review]
Super nice initiative. I like!

Question: If you want to fetch the translations in a non request scope, is that as easy as using `S3Translations.current`?

**Comment by mightyost** [comment]
Did we have a good strategy for using the lib in tests? I.e. returning something that is assertable and not brittle if someone updates the texts in S3?

**Comment by mightyost** [comment]
One last comment, I do not understand where `I18nQuotes` is coming from in your example in description?

**Comment by bystam** [comment]
> Super nice initiative. I like!
> 
> Question: If you want to fetch the translations in a non request scope, is that as easy as using `S3Translations.current`?

How do you mean? The `S3Translations` class is currently marked as `internal`, so you are not meant to interact with it directly. 

If you want to translate a string outside of a request scope, you can simply use `I18n` and set `hedvig.translations.warnOnContextlessUsage: false` to avoid the warnings being printed

> Did we have any good strategy for using the lib in tests? I.e. returning something that is assertable and not brittle if someone updates the texts in S3?

Yes, by setting
```yml
hedvig.translations:
  enabled: false # disable the fetching of translations
  missingTranslationMode: echo # make missing translations simply echo
``` 
for instance, you get the keys echoed back to you in a stable manner in the tests

> One last comment, I do not understand where I18nQuotes is coming from in your example in description?

That's a typo where I failed to update the PR description. Will fix.

File: translations/src/test/kotlin/com/hedvig/libs/translations/I18nTest.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    @AfterEach
+    fun tearDown() {
+        S3Translations.current = S3Translations.empty
+    }
+
+    @Test
+    fun `test missing keys`() {
+        assertThat(I18n.get("MY_KEY")).isEqualTo("MY_KEY")
+        assertThat(I18n.get("MY_KEY2", mapOf("name" to "Fredrik", "age" to 32))).isEqualTo("MY_KEY2{age=32, name=Fredrik}")
+    }
```
Comment (bystam) [inline translations/src/test/kotlin/com/hedvig/libs/translations/I18nTest.kt:28]: @mightyost Here's how the library behaves when testing :)

**Pull request merged**